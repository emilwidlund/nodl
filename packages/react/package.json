{
    "name": "@nodl/react",
    "version": "1.0.3",
    "description": "React implementation of the Nodl framework",
    "main": "build/index.js",
    "types": "build/index.d.ts",
    "author": "Emil Widlund",
    "license": "MIT",
    "private": false,
    "files": [
        "build/*"
    ],
    "scripts": {
        "build": "webpack --mode=production --devtool=inline-source-map",
        "build:prod": "webpack --mode=production --devtool=source-map",
        "generate:docs": "typedoc ./src/index.ts",
        "prepare": "yarn run build:prod",
        "test": "jest"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/emilwidlund/nodl.git"
    },
    "keywords": [],
    "bugs": {
        "url": "https://github.com/emilwidlund/nodl/issues"
    },
    "homepage": "https://github.com/emilwidlund/nodl#readme",
    "dependencies": {
        "@nodl/core": "^1.0.5",
        "@types/lodash": "^4.14.191",
        "@types/react": "^18.0.26",
        "lodash": "^4.17.21",
        "mobx": "^6.7.0",
        "mobx-react-lite": "^3.4.0",
        "react-draggable": "^4.4.5",
        "typescript": "^4.9.4"
    },
    "devDependencies": {
        "@emotion/jest": "^11.10.5",
        "@testing-library/jest-dom": "^5.16.5",
        "@testing-library/react": "^14.0.0",
        "@types/jest": "^29.2.5",
        "@types/react": "^18.0.28",
        "@types/react-dom": "^18.0.11",
        "@types/react-test-renderer": "^18.0.0",
        "@types/uuid": "^9.0.0",
        "copy-webpack-plugin": "^11.0.0",
        "jest": "^29.3.1",
        "jest-environment-jsdom": "^29.4.3",
        "ts-jest": "^29.0.3",
        "ts-loader": "^9.4.2",
        "typedoc": "^0.23.26",
        "webpack": "^5.75.0",
        "webpack-cli": "^5.0.1"
    },
    "peerDependencies": {
        "react": "^18.2.0",
        "react-dom": "^18.2.0"
    }
}
