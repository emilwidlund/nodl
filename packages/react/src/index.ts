import { enableStaticRendering } from 'mobx-react-lite';

if (typeof window !== 'undefined' && typeof window.document !== 'undefined') {
    const preconnectLink = document.createElement('link');
    preconnectLink.rel = 'preconnect';
    preconnectLink.href = 'https://rsms.me/';
    document.head.appendChild(preconnectLink);

    const stylesheetLink = document.createElement('link');
    stylesheetLink.rel = 'stylesheet';
    stylesheetLink.href = 'https://rsms.me/inter/inter.css';
    document.head.appendChild(stylesheetLink);
}

enableStaticRendering(typeof window === 'undefined');

export * from './components/Connection/Connection.types';
export * from './components/Canvas/Canvas.types';
export * from './components/Node/Node.types';
export * from './components/NodeWrapper/NodeWrapper';
export * from './components/NodeWrapper/NodeWrapper.types';
export * from './containers/Circuit/Circuit';
export * from './stores/CircuitStore/CircuitStore';
export * from './stores/CircuitStore/CircuitStore.types';
