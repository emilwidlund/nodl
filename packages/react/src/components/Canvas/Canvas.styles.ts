import { css } from '@emotion/react';

export const canvasWrapperStyles = css`
    position: relative;
    width: 100%;
    height: 100%;
    overflow: auto;
`;

export const canvasContentStyles = (size: { width: number; height: number }, zoomFactor: number) => css`
    position: absolute;
    top: 0;
    left: 0;
    width: ${size.width}px;
    height: ${size.height}px;
    background-image: radial-gradient(#434437 5%, #1c1e2a 5%);
    background-position: 0 0;
    background-size: 30px 30px;
    transform: scale(${zoomFactor});
    transform-origin: 0 0;
`;
