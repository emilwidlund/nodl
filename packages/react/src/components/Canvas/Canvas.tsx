/** @jsxImportSource @emotion/react */
import { observer } from 'mobx-react-lite';
import * as React from 'react';

import { canvasWrapperStyles, canvasContentStyles } from './Canvas.styles';
import { CanvasProps } from './Canvas.types';
import { StoreContext } from '../../stores/CircuitStore/CircuitStore';
import { fromCartesianPoint } from '../../utils/coordinates/coordinates';

export const Canvas = observer(
    React.forwardRef<HTMLDivElement, CanvasProps>(
        ({ children, size, className, onMouseMove, onClick, onMouseDown, onMouseUp }: CanvasProps, ref) => {
            const scrollRef = React.useRef<HTMLDivElement>(null);
            const { store } = React.useContext(StoreContext);

            React.useEffect(() => {
                if (scrollRef.current) {
                    const { x, y } = fromCartesianPoint(size.width, size.height, 0, 0);
                    const { x: offsetX, y: offsetY } = fromCartesianPoint(
                        scrollRef.current.clientWidth,
                        scrollRef.current.clientHeight,
                        0,
                        0
                    );

                    scrollRef.current.scrollTo({ left: x - offsetX, top: y - offsetY });
                }
            }, []);

            return (
                <div ref={scrollRef} css={canvasWrapperStyles} className={className}>
                    <div
                        ref={ref}
                        css={canvasContentStyles(size, store.zoomFactor)}
                        className="canvas"
                        children={children}
                        onMouseMove={onMouseMove}
                        onMouseDown={onMouseDown}
                        onMouseUp={onMouseUp}
                        onClick={onClick}
                    />
                </div>
            );
        }
    )
);
