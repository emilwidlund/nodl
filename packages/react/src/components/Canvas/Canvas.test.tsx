import { render, fireEvent } from '@testing-library/react';
import * as React from 'react';

import { Canvas } from './Canvas';
import { CanvasProps } from './Canvas.types';

const setup = (props?: Partial<CanvasProps>) => {
    return render(<Canvas size={{ width: 10000, height: 10000 }} {...props} />);
};

describe('Canvas', () => {
    it('should render correctly', () => {
        const result = setup();

        expect(result.asFragment()).toMatchSnapshot();
    });

    it('should scroll to cartesian midpoint on mount', () => {
        const scrollToSpy = jest.spyOn(Element.prototype, 'scrollTo');

        setup();

        expect(scrollToSpy).toHaveBeenNthCalledWith(1, { left: 10000 / 2, top: 10000 / 2 });
    });

    it('should invoke onMouseMove-handler', () => {
        const onMouseMove = jest.fn();

        const { container } = setup({ onMouseMove });
        const innerElement = container.getElementsByClassName('canvas')[0];

        fireEvent.mouseMove(innerElement);

        expect(onMouseMove).toHaveBeenCalled();
    });

    it('should invoke onMouseDown-handler', () => {
        const onMouseDown = jest.fn();

        const { container } = setup({ onMouseDown });

        const innerElement = container.getElementsByClassName('canvas')[0];
        fireEvent.mouseDown(innerElement);

        expect(onMouseDown).toHaveBeenCalledTimes(1);
    });

    it('should invoke onMouseUp-handler', () => {
        const onMouseUp = jest.fn();

        const { container } = setup({ onMouseUp });

        const innerElement = container.getElementsByClassName('canvas')[0];
        fireEvent.mouseUp(innerElement);

        expect(onMouseUp).toHaveBeenCalledTimes(1);
    });

    it('should invoke onClick-handler', () => {
        const onClick = jest.fn();

        const { container } = setup({ onClick });

        const innerElement = container.getElementsByClassName('canvas')[0];
        fireEvent.click(innerElement);

        expect(onClick).toHaveBeenCalledTimes(1);
    });
});
