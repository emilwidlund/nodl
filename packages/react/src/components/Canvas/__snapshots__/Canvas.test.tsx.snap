// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  width: 100%;
  height: 100%;
  overflow: auto;
}

.emotion-1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 10000px;
  height: 10000px;
  background-image: radial-gradient(#434437 5%, #1c1e2a 5%);
  -webkit-background-position: 0 0;
  background-position: 0 0;
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  -webkit-transform: scale(1.5);
  -moz-transform: scale(1.5);
  -ms-transform: scale(1.5);
  transform: scale(1.5);
  transform-origin: 0 0;
}

<div
    class="emotion-0"
  >
    <div
      class="canvas emotion-1"
    />
  </div>
</DocumentFragment>
`;
